### ISSUE TITLE
gc.freeze may be useful here?

### ISSUE THREAD

**Issue by fjsj-altana** [rainwoodman/sharedmem#19]
Hi, since this relies on copy-on-write to share variables, would it benefit from the `gc.freeze`? https://docs.python.org/3/library/gc.html#gc.freeze

More details: https://bugs.python.org/issue31558

**Comment by rainwoodman** [comment]
Thanks for the pointer.

We fork through multiprocessing.Process (except the experimental parallel.py).

From the reading I think gc.freeze() at enter + gc.unfreeze() at exit should have benefits when the parent has loads of python objects? We shall probably design a benchamark. Do you already have a case in mind?

**Comment by fjsj-altana** [comment]
> should have benefits when the parent has loads of python objects?

Yes, I think so, but haven't tested extensively. It seems freeze is used mostly in the context of web frameworks that use fork to start new workers.

> Do you already have a case in mind?

I need to perform parallel processing over a large dictionary. I don't want to broadcast parts of the dictionary to workers, because of the serialization costs. My current implementation is already constrained by these serialization costs. I want to try an alternative solution with CoW.

Does sharedmem do anything special to share Python objects other than using CoW? I just need a read-only copy of the dict in every worker.

**Comment by rainwoodman** [comment]
Thanks for the explanation.

In the Python data model, PyObject is always mutable due to the refcounting. Also complex objects like dict have their internal state that may mutate even from reading (at least some earlier versions of Python) .

sharedmem doesn't do anything special to share Python objects other than CoW. That's to say, if you Py-Incref a CoW object in the worker process, the entire page hosting the refcount of the object will be copied. This will happen regardless gc.freeze().
e.g, I suspect something as innocent as the following will make a copy of the page hosting the refcount of the object, which in turn is likely the entire Py_Object:

```
a = object()
fork()
b = a
```

On the other hand, this may be OK:
```
def call_a_function(b):
    pass

a = object()
fork()
call_a_function(a)
```

I think in general if you can somehow guarantee the size of PyObject struct is a small fraction of the memory footprint of the object then this overhead is negligible. For example if your dictionary items are large buffers, strings, or numpy arrays. Then you can guarantee the pages backing the buffer are immutable, and neglect the cost of CoW of the pages backing the PyObject.

**Comment by rainwoodman** [comment]
could you take a look if #20 does freeze() unfreeze() sensibly? I couldn't come up with a case to measure a difference of RSS usage before / after the change.

Perhaps a test case inspired by your use case? Here is what I used:
```
import numpy
import sharedmem
import time
import resource

from numpy.testing import (assert_equal, assert_array_equal,
    assert_almost_equal, assert_array_almost_equal, assert_, run_module_suite)

def test_expensive():
    """ An expensive workload for testing freeze.
    """
    c = dict([(i, str(i)) for i in range(1024*1024)])

    with sharedmem.MapReduce(np=4) as pool:

        def work(i):
            obj = c[i]
            if i > len(c) * 0.99:
                print(pool.local.rank, i, resource.getrusage(resource.RUSAGE_SELF).ru_maxrss)
            return len(obj)

        pool.map(work, range(len(c)))
```

**Comment by fjsj-altana** [comment]
> if your dictionary items are large buffers, strings, or numpy arrays. Then you can guarantee the pages backing the buffer are immutable, and neglect the cost of CoW of the pages backing the PyObject.

Yes, I think that's the key, so my use case probably won't benefit from forking.
I ended up greedily distributing the data across CPUs. It generates duplicates, but it gives enough memory reduction to fit per CPU.

I'll try to come up with a test case. I've found [this one from SO](https://stackoverflow.com/questions/49429368/how-to-solve-memory-issues-problems-while-multiprocessing-using-pool-map) where CoW helps, and maybe it can help even more with `gc.freeze`.
